### DISCUSSION TITLE
Avoiding upstream registry accesses and package immutability

### DISCUSSION THREAD

**Discussion by pboguslawski** [verdaccio/verdaccio#3457] [Q&A] [answered]
When fetching package via verdaccio

```
NPM_CONFIG_REGISTRY=https://verdaccio.mynet npm install test@1.2.4
```

verdaccio calls upstream registry (https://registry.npmjs.org/) and saves package it in its storage folder. Then another request

```
NPM_CONFIG_REGISTRY=https://verdaccio.mynet npm install test@1.2.4
```

makes verdaccio calling upstream registry (https://registry.npmjs.org/) again even if it already has this package in its storage.

Is it possible to configure  verdaccio to avoid calling upstream if package is already available in local storage (i.e. for privacy reasons)?

Does/may verdaccio guarantee package immutability which is required for deterministic builds, i.e. when package version was already cached, should not be fetched again from upstream registry even if upstream changed the same version (similar to how [goproxy](https://jfrog.com/blog/why-goproxy-matters-and-which-to-pick/) works for go modules)?

**Comment by juanpicado** [comment, marked as answer]
Hi, I didn't notice your question until now.

> Is it possible to configure verdaccio to avoid calling upstream if package is already available in local storage (i.e. for privacy reasons)?

You can do it by increasing `maxage` for year or something longer to avoid the metadata mutates https://verdaccio.org/docs/uplinks/#configuration
 
> Does/may verdaccio guarantee package immutability which is required for deterministic builds, i.e. when package version was already cached, should not be fetched again from upstream registry even if upstream changed the same version (similar to how [goproxy]( ) works for go modules)?

immutability only applies to tarball, tarballs are Neve downloaded twice, only metadata mutates. I don't think this issue applies to Node.js registries.

**Comment by pboguslawski** [comment]
> immutability only applies to tarball, tarballs are Neve downloaded twice, only metadata mutates. 

Thank you. More questions:

(1)
After caching packgage in Verdaccio with

```NPM_CONFIG_REGISTRY=https://verdaccio.mynet npm install test@1.2.4```

and default `maxage` will pass, how will Verdaccio react on another

```NPM_CONFIG_REGISTRY=https://verdaccio.mynet npm install test@1.2.4```

if `test@1.2.4` will change its contents to something different (new content with matching new checksum/signature, i.e. repository compromise) in upstream repository?

(2)
After caching package in Verdaccio with

```NPM_CONFIG_REGISTRY=https://verdaccio.mynet npm install test@latest```

and setting `maxage=5270400m`  how will Verdaccio react on another 

```NPM_CONFIG_REGISTRY=https://verdaccio.mynet npm install test@latest```

when `test@latest` will change its contents to something different (new content with matching new checksum/signature, i.e. package upgrade) in upstream repository and cache is still valid (10y)?

(3)
Does `m` mean `minute` or `month`? If minute - is it possible to specify `maxage=10y` instead of `maxage=5270400m`? 

(4)
Is it possible in Verdaccio:
* if asked to serve specified `X.Y.Z` version that is already downloaded locally - serve it from local storage forever (don't use `maxage` here) and don't ask upstream for anything connected with this `X.Y.Z ` version (for privacy reasons),
* if asked to serve tag (i.e. `latest`) ask upstream server (skip asking only if `maxage` not passed after previous same tag query) and update local tag if changed by upstream (to allow auto updating packages).
